lib/src/code_field/code_controller.dart

```diff
@@ -316,20 +316,32 @@ class CodeController extends TextEditingController {
         return;
       }
 
-      _updateLastCodeIfChanged(editResult.fullTextAfter);
+      final isTabReplacingEnabled = modifiers.any(
+        (element) => element is TabModifier,
+      );
+      final textAfterWithoutTabs =
+          editResult.fullTextAfter.replaceAll('\t', ' ' * params.tabSpaces);
 
-      if (newValue.text != _lastCode.visibleText) {
-        // Manually typed in a text that has become a hidden range.
-        newValue = newValue.replacedText(_lastCode.visibleText);
+      if (isTabReplacingEnabled) {
+        _updateLastCodeIfChanged(textAfterWithoutTabs);
+      } else {
+        _updateLastCodeIfChanged(editResult.fullTextAfter);
       }
 
+      newValue = _updateSelectionPosition(
+        newValue,
+        editResult,
+        textAfterWithoutTabs,
+        isTabReplacingEnabled,
+      );
       // Uncomment this to see the hidden text in the console
       // as you change the visible text.
       //print('\n\n${_lastCode.text}');
     }
 
-    bool hasTextChanged = newValue.text != super.value.text;
-    bool hasSelectionChanged = newValue.selection != super.value.selection;
+    final bool hasTextChanged = newValue.text != super.value.text;
+    final bool hasSelectionChanged =
+        newValue.selection != super.value.selection;
```

Comment:
I guess these were kept non-final to minimize the diff with the original editor. Please revert this.